lib/type_check/external.ex

```diff
@@ -0,0 +1,198 @@
+      iex> type
+      #TypeCheck.Type< (number() -> number()) >
+      iex> {:ok, type} = fetch_spec(Atom, :to_string, 1)
+      iex> type
+      #TypeCheck.Type< (atom() -> binary()) >
+      iex> fetch_spec(Kernel, :non_existent, 1)
+      {:error, "cannot find spec for function"}
+  """
+  @spec fetch_spec(module(), atom(), arity()) :: TypeCheck.Type.t()
+  def fetch_spec(module, function, arity) do
+    case Parser.fetch_spec(module, function, arity) do
+      {:ok, spec} -> {:ok, Parser.convert(spec)}
+      {:error, err} -> {:error, err}
+    end
+  end
+
+  @doc """
+  Extract TypeCheck type from the regular Elixir (or Erlang) `@type` with the given name.
+
+  To fetch a generic type, you must pass a list of types to be placed instead of
+  generic variables. If you don't know these types, just punch in `any()`
+  as many times as you need. For example, to fetch the type of `Range.t(left, right)`,
+  you need to pass `[number(), number()]` or `[any(), any()]`.
+
+  ## Examples
+
+  Fetching a regular type:
+
+      iex> import TypeCheck.External
+      iex> {:ok, type} = fetch_type(String, :t)
+      iex> type
+      #TypeCheck.Type< binary() >
+
+  Fetching a generic type:
+
+      iex> import TypeCheck.External
+      iex> import TypeCheck.Builtin
+      iex> {:ok, type} = fetch_type(:elixir, :keyword, [number()])
+      iex> type
+      #TypeCheck.Type< list({atom(), number()}) >
+      iex> {:ok, type} = fetch_type(Range, :t, [integer(), integer()])
+      iex> type
+      #TypeCheck.Type< %Range{first: integer(), last: integer(), step: pos_integer() | neg_integer()} >
+
+  Fetching non-existent type causes an error:
+
+      iex> import TypeCheck.External
+      iex> fetch_type(String, :non_existent)
+      {:error, "cannot find type with the given name"}
+  """
+  @spec fetch_type(module(), atom(), [TypeCheck.Type.t()]) :: TypeCheck.Type.t()
+  def fetch_type(module, type, var_types \\ []) do
+    arity = length(var_types)
+
+    case Parser.fetch_type(module, type, arity) do
+      {:ok, type, var_names} ->
+        vars = Enum.zip(var_names, var_types) |> Map.new()
+        ctx = %{Parser.Context.default() | vars: vars, module: module}
+        {:ok, Parser.convert(type, ctx)}
+
+      {:error, err} ->
+        {:error, err}
+    end
+  end
+
+  @doc """
+  The same as `Kernel.apply/3` but ensures that values conform the given type.
+
+  The first argument must be a function type of the function to be called.
+
+  In case of type error, raises an exception.
+
+  ## Examples
+
+      iex> alias TypeCheck.Builtin, as: B
+      iex> type = B.function([B.number], B.number)
+      iex> TypeCheck.External.apply!(type, Kernel, :abs, [-13])
+      13
+      iex> TypeCheck.External.apply!(type, Kernel, :abs, ["hello"])
+      ** (TypeCheck.TypeError) At lib/type_check.ex:279:
+          `"hello"` is not a number.
+
+  """
+  @spec apply!(TypeCheck.Type.t(), module(), atom(), list()) :: any()
+  def apply!(type, module, function, args) do
```

Comment:
If you want, you can change this to re-use `apply`, by writing:

```
case apply(type, module, function, args) do
  {:ok, result} -> result
  {:error, exception} -> raise exception
end
```

Reply:
I didn't want to mess with the stacktrace, TBH. Otherwise, yes, just one function call is actually different.